primitives/numeric/src/lib.rs

```diff
@@ -0,0 +1,531 @@
+        where
+            D: ::serde::Deserializer<'de>,
+        {
+            rust_decimal::serde::str::deserialize(deserializer).and_then(|inner| {
+                inner
+                    .is_sign_positive()
+                    .then_some(Self { inner })
+                    .ok_or_else(|| D::Error::custom("number must be non negative"))
+            })
+        }
+    }
+}
+
+mod scale_ {
+    use parity_scale_codec::{Decode, Encode};
+
+    use super::*;
+
+    #[derive(Encode, Decode)]
+    // Use compact encoding for efficiency, for integer numbers scale takes only one byte
+    struct NumericScaleHelper {
+        #[codec(compact)]
+        mantissa: u128,
+        #[codec(compact)]
+        scale: u32,
+    }
+
+    impl Encode for Numeric {
+        fn encode(&self) -> Vec<u8> {
+            NumericScaleHelper {
+                mantissa: self.mantissa(),
+                scale: self.scale(),
+            }
+            .encode()
+        }
+    }
+
+    impl Decode for Numeric {
+        fn decode<I: parity_scale_codec::Input>(
+            input: &mut I,
+        ) -> Result<Self, parity_scale_codec::Error> {
+            let NumericScaleHelper { mantissa, scale } = NumericScaleHelper::decode(input)?;
+            match Numeric::try_new(mantissa, scale) {
+                Ok(numeric) => Ok(numeric),
+                Err(NumericError::MantissaTooLarge) => {
+                    Err("error decoding numeric: mantissa too large".into())
+                }
+                Err(NumericError::ScaleTooLarge) => {
+                    Err("error decoding numeric: scale too large".into())
+                }
+                // Not possible to get malformed or negative value from mantissa and scale
+                Err(NumericError::Malformed | NumericError::Negative) => unreachable!(),
+            }
+        }
+    }
+}
+
+mod schema_ {
+    use iroha_schema::{
+        Declaration, Ident, IntoSchema, MetaMap, Metadata, NamedFieldsMeta, TypeId,
+    };
+
+    use super::*;
+
+    impl TypeId for Numeric {
+        fn id() -> Ident {
+            "Numeric".to_string()
+        }
+    }
+
+    impl IntoSchema for Numeric {
+        fn type_name() -> Ident {
+            "Numeric".to_string()
+        }
+
+        fn update_schema_map(metamap: &mut MetaMap) {
+            if !metamap.contains_key::<Self>() {
+                if !metamap.contains_key::<u128>() {
+                    <u128 as iroha_schema::IntoSchema>::update_schema_map(metamap);
+                }
+                if !metamap.contains_key::<u32>() {
+                    <u32 as iroha_schema::IntoSchema>::update_schema_map(metamap);
+                }
+
+                metamap.insert::<Self>(Metadata::Struct(NamedFieldsMeta {
+                    declarations: vec![
+                        Declaration {
+                            name: "mantissa".to_string(),
+                            ty: core::any::TypeId::of::<u128>(),
```

Comment:
Shouldn't this be `Compact<u128>`, since the `NumericScaleHelper`'s fields are annotated with `#[scale(compact)]`?